### PR TITLE
Pattoo installation

### PR DESCRIPTION
**Pull request by DangaRanga**
Changes to installation:
 
 - Created configure class to allow the default server and pattoo directory configurations to be more accessible and to allow for a more universal configuration
 -  Modified the pattoo_config method to configure the server and pattoo directories
 - Made systemd installation more universal

### CODE REVIEW

File: pattoo_shared/installation/configure.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+# Dependendices
+import yaml
+
+# Import project libraries
+from pattoo_shared import files, configuration, log
+from pattoo_shared.installation import shared
+
+
+class Configure():
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:16]: We are going to have to review this. Many of your methods don't use class attributes and should therefore be functions.

File: pattoo_shared/installation/configure.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            # If the user specified does not exist, they get created
+            if not self.user_exists(user_name):
+                shared.run_script(
+                    'useradd -d {1} -s {2} -g {0} {0}'.format(
+                        user_name, directory, shell), verbose)
+        else:
+            # Die if not root
+            shared.log('You are currently not running the script as root')
+
+    def group_exists(self, group_name):
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:104]: This should be a function not a method. It uses no instance variables.

File: pattoo_shared/installation/configure.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            True if the group exists and False if it does not
+        """
+        try:
+            # Gets group name
+            grp.getgrnam(group_name)
+            return True
+        except KeyError:
+            return False
+
+    def user_exists(self, user_name):
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:120]: This should be a function not a method. It uses no instance variables.

File: pattoo_shared/installation/configure.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+                        # Recursively set file ownership to pattoo user
+                        # and group
+                        shared.chown(full_directory)
+
+        # Write file
+        with open(config_file, 'w') as f_handle:
+            yaml.dump(config, f_handle, default_flow_style=False)
+
+    def read_config(self, filepath, default_config):
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:193]: This should be a function not a method. It uses no instance variables.

File: pattoo_shared/installation/configure.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+        """
+        try:
+            # Gets user name
+            pwd.getpwnam(user_name)
+            return True
+        except KeyError:
+            return False
+
+    def pattoo_config(
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:137]: This should be a function not a method. It uses no instance variables.

File: pattoo_shared/installation/configure.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+            with open(filepath, 'r') as f_handle:
+                yaml_string = (
+                    '{}\n{}'.format(default_config_string, f_handle.read()))
+                config = yaml.safe_load(yaml_string)
+        else:
+            config = default_config
+
+        return config
+
+    def secondary_key_check(self, config, primary, secondaries):
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:218]: This should be a function not a method. It uses no instance variables.

File: pattoo_shared/installation/configure.py (hunk 5 above)
Comment (palisadoes) [reply pattoo_shared/installation/configure.py:137]: At least, only uses two, but why also pass arguments?

Comment (palisadoes) [reply pattoo_shared/installation/configure.py:137]: Ideally, you should pass two variables. The default dict, and the dict read in from the configuration directory. The function should return the result.

File: pattoo_shared/installation/configure.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        """
+        # Check keys
+        for key in secondaries:
+            if key not in config[primary]:
+                log_message = ('''\
+Configuration file's "{}" section does not have a "{}" sub-section. \
+Please fix.'''.format(primary, key))
+                log.log2die_safe(1057, log_message)
+
+    def check_config(self):
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:238]: Ideally this function should have the content of the configuration files, the primary keys and secondary keys as arguments. It doesn't need to be a method. It won't scale as written for other agents.

File: pattoo_shared/installation/configure.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+            'log_level', 'log_directory', 'cache_directory',
+            'daemon_directory']
+        self.secondary_key_check(config, 'pattoo', secondaries)
+        secondaries = ['ip_address', 'ip_bind_port']
+        self.secondary_key_check(config, 'pattoo_agent_api', secondaries)
+
+        # Print Status
+        print('OK: Configuration parameter check passed.')
+
+    def run_configure(
```
Comment (palisadoes) [inline pattoo_shared/installation/configure.py:278]: This should be left in pattoo or the agent that is being configured.

File: pattoo_shared/installation/configure.py (hunk 5 above)
Comment (DangaRanga) [reply pattoo_shared/installation/configure.py:137]: I just went back through the code and realized that read_config goes through the default config file and returns the result based on the dictionary passed in, so this function only makes the directories and creates the configuration file.
Would it still be necessary to return the result?

**Pull request merged**